spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+   * talking to Stackdriver.
+   */
+  private static GoogleCredential loadCredential(
+      HttpTransport transport, JsonFactory factory, String credentialsPath)
+      throws IOException {
+    GoogleCredential credential;
+    if (!credentialsPath.isEmpty()) {
+      FileInputStream stream = new FileInputStream(credentialsPath);
+      credential = GoogleCredential.fromStream(stream, transport, factory)
+                   .createScoped(
+                         Collections.singleton(MonitoringScopes.MONITORING));
+      log.info("Loaded credentials from from {}", credentialsPath);
+    } else {
+      log.info("spectator.stackdriver.monitoring.enabled without"
+               + " spectator.stackdriver.credentialsPath. "
+               + " Using default application credentials.");
+      credential = GoogleCredential.getApplicationDefault();
+    }
+    return credential;
+  }
+
+  /**
+   * Constructs a writer instance.
+   */
+  public StackdriverWriter(
+      Monitoring monitoring,
+      String customTypeNamespace,
+      String projectName, String applicationName, String applicationVersion,
+      MeasurementFilter measurementFilter) {
+    if (projectName.isEmpty()) {
+      throw new IllegalStateException(
+          "You need to configure a stackdriver.projectName.");
+    }
+    this.service = monitoring;
+    this.customTypePrefix = String.format("custom.googleapis.com/%s/",
+                                          customTypeNamespace);
+    this.projectResourceName = "projects/" + projectName;
+    this.applicationName = applicationName;
+    this.measurementFilter = measurementFilter;
+
+    UUID uuid = UUID.randomUUID();
+    byte[] uuidBytes = new byte[16];
+    addLong(uuidBytes, 0, uuid.getLeastSignificantBits());
+    addLong(uuidBytes, 8, uuid.getMostSignificantBits());
+    setInstanceId(java.util.Base64.getEncoder().encodeToString(uuidBytes));
+  }
+
+  /**
+   * Helper function to encode a value into buffer when constructing a UUID.
+   */
+  private void addLong(byte[] buffer, int offset, long value) {
+    for (int i = 0; i < 8; ++i) {
+      buffer[i + offset] = (byte)(value >> (8 * i) & 0xff);
+    }
+  }
+
+  /**
+   * [Re]bind the instanceId used to identify the TimeSeries data coming from
+   * this particular instance.
+   */
+  public void setInstanceId(String id) {
+    log.info("Configuring {}={}", INSTANCE_LABEL, id);
+    instanceId = id;
+  }
+    
+  /**
+   * Only useful for testing.
+   */
+  public String getInstanceId() {
+    return instanceId;
+  }
+
+  /**
+   * Lookup all the pre-existing custom metrics for spinnaker so that we
+   * know if we need to create new metric descriptors or not.
+   */
+  public void initKnownDescriptors()
+      throws HttpResponseException, IOException {
+    ListMetricDescriptorsResponse response =
+      service.projects().metricDescriptors().list(this.projectResourceName)
+      .execute();
```

Comment:
How big is this list expected to be? For a number of our apps I would expect this to be quite large. If we spin up a bunch of instances at the same time, then will it be an issue if all are trying to list all of the descriptors?